Tutorial - How to add popups to your React app with PostHog (#5488)
PostHog/posthog.com: contents/tutorials/react-popups.md

---

> title: How to add popups to your React app with PostHog

Reviewer: Q: Do we think it's worth getting Feature Flags / JSON into the title?

date: 2023-03-17
author: ["ian-vanagas"]
showTitle: true
sidebar: Docs
featuredImage: ../images/tutorials/banners/tutorial-3.png
topics: ['feature flags', 'experimentation']
---

> Popups are a way to highlight features in your app. This tutorial shows how to add them to a React app and control them using feature flags and JSON payloads. The use of feature flag payloads enables easy customization and control of popups.

Reviewer: ```suggestion
Popups are a way to highlight features in your app. This tutorial shows how to add them to a React app and control them using feature flags and JSON payloads. Feature flag payloads are an ideal tool for this task because they enable you to send arbitrary data, which you can use to control popups without needing to redeploy. 
```

## Creating a React app

We need a React app to start. Make sure you [installed Node](https://nodejs.dev/en/learn/how-to-install-nodejs/), then use the `create-react-app` command in your terminal to create an app named "popups":

```bash
npx create-react-app popups
```

> Next, go into the newly created popups folder and run the app to make sure everyone installed correctly.

Reviewer: ```suggestion
Next, go into the newly created popups folder and run the app to make sure everything installed correctly.
```

## Adding the code for the popup

Next, add the control that displays the popups and control when and where they display. 

Clone the code from [PostHog prompts repository](https://github.com/PostHog/posthog-prompts). 

```bash
git clone https://github.com/PostHog/posthog-prompts
```
In the newly cloned `posthog-prompts` folder, go to `react/src`, then copy the `Popup.js` `Popup.css` and `popup-location.js` files over to our `popups/src` folder.

> You could also create the files yourself and manually copy the code from the [popup](https://github.com/PostHog/posthog-prompts/blob/main/react/src/Popup.js), [CSS](https://github.com/PostHog/posthog-prompts/blob/main/react/src/Popup.css), and [location](https://github.com/PostHog/posthog-prompts/blob/main/react/src/popup-location.js) files.

In `index.js`, import the popup code you just added and add the `<Popup />` component to the app.

```js
// src/index.js
// ...
import { Popup } from './Popup'

//...

const root = ReactDOM.createRoot(document.getElementById('root'));
root.render(
  <React.StrictMode>
    <PostHogProvider client={posthog}>
      <App />
      <Popup />
    </PostHogProvider>
  </React.StrictMode>
);
```

> After doing this, you’re ready to set up the feature flag to control your popup.

Reviewer: ```suggestion
After this, you’re ready to set up the feature flag to control your popup.
```

## Create your popup feature flag

The next step is to create a feature flag with a payload to control the content. In PostHog, go to the feature flag tab and click the "New feature flag" button.

1. Name your flag something starting with `popup-`
2. Add an extra filter to exclude users with the property "`$popup-{unique name}` is not set" so that the popup is only shown once for each user
3. Set the rollout to 100%
4. Add your content in JSON as the payload like:

```js
{
    "title": "Learn more about PostHog",
    "body": "PostHog is an all in one suite of product and data tools",
    "primaryButtonText": "Go to PostHog",
    "primaryButtonURL": "/",
    "secondaryButtonText": "Got it",
    "location": "absolute-bottom-right"
}
```

> Save the flag, head back to your app, and you see a popup in the bottom right corner.

Reviewer: ```suggestion
Save the flag, head back to your app, and you should see a popup in the bottom right corner.
```
